artichoke-backend/src/extn/core/string/trampoline.rs

```diff
@@ -1478,54 +1478,113 @@ pub fn to_f(interp: &mut Artichoke, mut value: Value) -> Result<Value, Error> {
+        Some(b'+') => {
+            slice = &slice[1..];
+            1
         }
-    } else {
-        10_u32
+        Some(b'-') => {
+            slice = &slice[1..];
+            -1
+        }
+        _ => 1,
     };
-    let mut slice = s.as_slice();
-    let mut squeezed = false;
-    // squeeze preceding zeros.
-    while let Some(&b'0') = slice.first() {
-        slice = &slice[1..];
-        squeezed = true;
-    }
-    // Trim leading literal specifier but only if there was a leading 0.
-    if squeezed
-        && matches!(
-            (base, slice.first().copied()),
-            (2, Some(b'b' | b'B')) | (8, Some(b'o' | b'O')) | (10, Some(b'd' | b'D')) | (16, Some(b'x' | b'X'))
-        )
-    {
-        slice = &slice[1..];
-    }
 
-    if slice.is_empty() {
+    let base = base.map_or(Ok(10), |b| implicitly_convert_to_int(interp, b))?;
+    let base = match base {
+        x if x < 0 || x == 1 || x > 36 => return Err(ArgumentError::from(format!("invalid radix {base}")).into()),
+        0 => {
+            // Infer base size from prefix
+            if slice.len() < 2 || slice[0] != b'0' {
+                10
+            } else {
+                match &slice[1] {
+                    b'b' | b'B' => {
+                        slice = &slice[2..];
+                        2
+                    }
+                    b'o' | b'O' => {
+                        slice = &slice[2..];
+                        8
+                    }
+                    b'd' | b'D' => {
+                        slice = &slice[2..];
+                        10
+                    }
+                    b'x' | b'X' => {
+                        slice = &slice[2..];
+                        16
+                    }
+                    _ => {
+                        // Numbers that start with 0 are assumed to be octal
+                        slice = &slice[1..];
+                        8
+                    }
+                }
+            }
+        }
+        x => {
+            // Trim leading literal specifier if it exists
+            if slice.len() >= 2
+                && matches!(
+                    (x, &slice[0..2]),
+                    (2, b"0b" | b"0B") | (8, b"0o" | b"0O") | (10, b"0d" | b"0D") | (16, b"0x" | b"0X")
+                )
+            {
+                slice = &slice[2..];
+            };
+
+            // This can only be 2-36 inclusive in this branch, so unwrap is safe
+            u32::try_from(x).unwrap()
+        }
+    };
+
+    // Check string doesn't start with any special characters
+    //  '_' is invalid because they are stripped out elsewhere in the string, but cannot start a number
+    //  '+' and '-' invalid because we already have the sign prior to the prefix, but they are accepted
+    //  at the begining of a string by str_from_radix, and double sign doesn't make sense
+    if slice.first() == Some(&b'_') || slice.first() == Some(&b'+') || slice.first() == Some(&b'-') {
```

Comment:
I like the way this can be cleaned up with `matches!`:

```suggestion
    if matches!(slice.first(), Some(&b'_') | Some(&b'+') | Some(&b'-')) {
```